File: app/CicadaApp/Sources/CicadaApp/Models/ProjectState.swift
import Foundation

/// G141 §6.2 — a project's derived state on one day, the Swift twin of `api/services/project_state.py` (R-PP4,
/// R-PJB17). Both run ONE table, `api/tests/fixtures/timeline_state.json`, so the app and an agent (`cicada_project`)
/// never disagree about whether a thread is quiet or a milestone overdue. `today` is an argument: nothing here reads a
/// clock, and nothing here is stored (R-PJ7). Keep it line for line with the Python — a rule changed on one side
/// only fails `ProjectStateTests`.
enum ProjectState {
    static let quietFloorDays = 14        // R-PJ13
    static let quietMultiplier = 2.0      // R-PJ13: 2×, not derived-first's 3×
    static let followupFloorDays = 21     // R-PJ13
    static let overdueAskDays = 3         // §9: asked about once 3 days overdue
    static let gapWindowDays = 180        // R-PJB6: ending at the last moment day
    static let quietSectionDays = 90      // §6.2: Quiet ≤ 90 days, Resting beyond
    static let restingStatuses: Set<String> = ["decaying", "archived"]
    static let counted: Set<String> = ["planned", "done", "missed", "passed-no-word"]   // R-PJ11: dropped never counts

    enum Section: String, Decodable, Sendable { case inMotion, quiet, resting }

    /// A payload's absolute fields (`input_from_timeline`), or the fixture's hand-written ones.
    struct Input: Decodable, Sendable {
        var status: String
        var momentDays: [String]
        var lastMomentDay: String?
        var medianGapDays: Double?
        var openThreads: [ProjectOpenThread]
        var milestones: [ProjectMilestone]

        init(status: String = "active", momentDays: [String] = [], lastMomentDay: String? = nil,
             medianGapDays: Double? = nil, openThreads: [ProjectOpenThread] = [], milestones: [ProjectMilestone] = []) {
            self.status = status
            self.momentDays = momentDays
            self.lastMomentDay = lastMomentDay
            self.medianGapDays = medianGapDays
            self.openThreads = openThreads
            self.milestones = milestones
        }

        init(_ row: ProjectRow) {
            self.init(status: row.status, lastMomentDay: row.lastMomentDay, medianGapDays: row.medianGapDays,
                      openThreads: row.openThreads, milestones: row.milestones)
        }

        init(_ t: ProjectTimeline) {
            self.init(status: t.project.status, momentDays: t.momentDays, lastMomentDay: t.lastMomentDay,
                      medianGapDays: t.medianGapDays, openThreads: t.now.threads, milestones: t.milestones)
        }

        enum CodingKeys: String, CodingKey { case status, momentDays, lastMomentDay, medianGapDays, openThreads, milestones }

        init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            status = c.lenient(.status, "active")
            momentDays = c.lenient(.momentDays, [])
            lastMomentDay = c.lenient(.lastMomentDay)
            medianGapDays = c.lenient(.medianGapDays)
            openThreads = c.lenient(.openThreads, [])
            milestones = c.lenient(.milestones, [])
        }
    }

    struct ThreadState: Decodable, Equatable, Sendable {
        let claimId: String
        let quietDays: Int?
        let followupEligible: Bool
    }

    struct MilestoneState: Decodable, Equatable, Sendable {
        let slug: String
        /// `upcoming | overdue | someday | done | missed | dropped | passed-no-word`.
        let state: String
        /// Days to the target (planned), or done − target (done: < 0 early, > 0 late).
        let days: Int?
        let moved: Bool
        let followupEligible: Bool
    }

    struct Output: Decodable, Equatable, Sendable {
        let medianGapDays: Double?
        let quietThreshold: Int
        let section: Section
        let threads: [ThreadState]
        let milestones: [MilestoneState]
        let progress: ProjectProgress
        let next: String?
        let planned: Bool

        func thread(_ claimId: String) -> ThreadState? { threads.first { $0.claimId == claimId } }
        func milestone(_ slug: String) -> MilestoneState? { milestones.first { $0.slug == slug } }

        /// R-PP4 — quiet past Q, the section's own rule (`idle <= q` is in motion).
        func isQuiet(_ claimId: String) -> Bool { (thread(claimId)?.quietDays ?? 0) > quietThreshold }
    }

    /// The median gap between distinct moment days in the 180 days ending at the LAST one — data-anchored, so the
    /// server can serve it without today (R-PJB6).
    static func medianGap(_ days: [String]) -> Double? {
        let ds = Array(Set(days.compactMap { ISODay($0) })).sorted()
        guard let last = ds.last, ds.count >= 2 else { return nil }
        let window = ds.filter { last - $0 <= gapWindowDays }
        let gaps = zip(window, window.dropFirst()).map { $1 - $0 }.sorted()
        guard !gaps.isEmpty else { return nil }
        let mid = gaps.count / 2
        return gaps.count % 2 == 1 ? Double(gaps[mid]) : Double(gaps[mid - 1] + gaps[mid]) / 2
    }

    /// Q = max(14, 2 × median gap), rounded half-up.
    static func quietThreshold(_ gap: Double?) -> Int {
        guard let gap else { return quietFloorDays }
        return max(quietFloorDays, Int((quietMultiplier * gap + 0.5).rounded(.down)))
    }

    static func milestoneState(_ m: ProjectMilestone, today: ISODay) -> MilestoneState {
        let target = ISODay(m.target)
        let doneOn = ISODay(m.doneOn)
        switch m.status {
        case "planned":
            guard let target else {
                return MilestoneState(slug: m.slug, state: "someday", days: nil, moved: m.moved, followupEligible: false)
            }
            return MilestoneState(slug: m.slug, state: target >= today ? "upcoming" : "overdue", days: target - today,
                                  moved: m.moved, followupEligible: today - target >= overdueAskDays)
        case "done":
            let days: Int? = if let target, let doneOn { doneOn - target } else { nil }
            return MilestoneState(slug: m.slug, state: "done", days: days, moved: m.moved, followupEligible: false)
        case "missed", "dropped", "passed-no-word":
            return MilestoneState(slug: m.slug, state: m.status, days: nil, moved: m.moved, followupEligible: false)
        default:
            return MilestoneState(slug: m.slug, state: "someday", days: nil, moved: m.moved, followupEligible: false)
        }
    }

    static func progress(_ milestones: [ProjectMilestone]) -> ProjectProgress {
        let goals = milestones.filter { $0.source != "expectedEnd" && counted.contains($0.status) }
        return ProjectProgress(done: goals.filter { $0.status == "done" }.count, total: goals.count)
    }

    /// The open planned milestone with the earliest target, undated last — upcoming OR overdue, found without today.
    static func nextSlug(_ milestones: [ProjectMilestone]) -> String? {
        milestones
            .filter { $0.status == "planned" && $0.source != "expectedEnd" }
            .sorted { a, b in
                let ka = (a.target == nil ? 1 : 0, a.target ?? "", a.slug)
                let kb = (b.target == nil ? 1 : 0, b.target ?? "", b.slug)
                return ka < kb
            }
            .first?.slug
    }

    static func state(_ input: Input, today: ISODay) -> Output {
        var gap = input.medianGapDays
        if gap == nil, !input.momentDays.isEmpty { gap = medianGap(input.momentDays) }
        let q = quietThreshold(gap)
        let last = ISODay(input.lastMomentDay) ?? input.momentDays.compactMap { ISODay($0) }.max()
        let idle = last.map { today - $0 }
        let section: Section
        if restingStatuses.contains(input.status) || idle == nil || (idle ?? 0) > quietSectionDays {
            section = .resting
        } else if let idle, idle <= q {
            section = .inMotion
        } else {
            section = .quiet
        }
        let threads = input.openThreads.map { t -> ThreadState in
            let heard = ISODay(t.lastHeard) ?? ISODay(t.since)
            let quietDays = heard.map { today - $0 }
            return ThreadState(claimId: t.claimId, quietDays: quietDays,
                               followupEligible: quietDays.map { $0 >= max(followupFloorDays, q) } ?? false)
        }
        return Output(
            medianGapDays: gap, quietThreshold: q, section: section, threads: threads,
            milestones: input.milestones.filter { $0.source != "expectedEnd" }.map { milestoneState($0, today: today) },
            progress: progress(input.milestones), next: nextSlug(input.milestones), planned: !input.milestones.isEmpty)
    }
}

File: api/services/project_state.py
"""G141 §6.2 — a project's derived state on one day: one pure function.

Python here (for `cicada_project` and the follow-up proposer) and Swift in
PJ-5 (`ProjectState.swift`) run ONE fixture, `api/tests/fixtures/timeline_state.json`
— the QuickMatch/text_fold precedent (R-PJB17): two languages, one table, so the
app and an agent never disagree about whether a thread is quiet. `today` is an
argument; nothing here reads a clock and nothing is stored (R-PJ7).
"""
from __future__ import annotations

import math
from datetime import date
from statistics import median

QUIET_FLOOR_DAYS = 14          # R-PJ13
QUIET_MULTIPLIER = 2           # R-PJ13: 2×, not derived-first's 3×
FOLLOWUP_FLOOR_DAYS = 21       # R-PJ13
OVERDUE_ASK_DAYS = 3           # §9: a milestone is asked about once 3 days overdue
GAP_WINDOW_DAYS = 180          # R-PJB6: ending at the last moment day
QUIET_SECTION_DAYS = 90        # §6.2: Quiet ≤ 90 days, Resting beyond
RESTING_STATUSES = frozenset({"decaying", "archived"})
COUNTED = frozenset({"planned", "done", "missed", "passed-no-word"})   # R-PJ11: dropped never counts


def _d(value) -> date | None:
    try:
        return date.fromisoformat(str(value)[:10])
    except (TypeError, ValueError):
        return None


def median_gap(days) -> float | None:
    """Median gap between distinct moment days in the 180 days ending at the
    LAST one — data-anchored, so a server can serve it without today (R-PJB6)."""
    ds = sorted({x for x in (_d(v) for v in days or []) if x})
    if len(ds) < 2:
        return None
    ds = [x for x in ds if (ds[-1] - x).days <= GAP_WINDOW_DAYS]
    gaps = [(b - a).days for a, b in zip(ds, ds[1:])]
    return float(median(gaps)) if gaps else None


def quiet_threshold(median_gap_days: float | None) -> int:
    """Q = max(14, 2 × median gap), rounded half-up — Swift's `.rounded()`."""
    if median_gap_days is None:
        return QUIET_FLOOR_DAYS
    return max(QUIET_FLOOR_DAYS, int(math.floor(QUIET_MULTIPLIER * median_gap_days + 0.5)))


def milestone_state(m: dict, today: date) -> dict:
    status, target, done_on = m.get("status"), _d(m.get("target")), _d(m.get("doneOn"))
    out: dict = {"slug": m.get("slug"), "moved": bool(m.get("moved")), "days": None, "followupEligible": False}
    if status == "planned":
        if target is None:
            out["state"] = "someday"
        else:
            out["days"] = (target - today).days
            out["state"] = "upcoming" if target >= today else "overdue"
            out["followupEligible"] = (today - target).days >= OVERDUE_ASK_DAYS
    elif status == "done":
        out["state"] = "done"
        if target and done_on:
            out["days"] = (done_on - target).days   # < 0 early, 0 on time, > 0 late
    elif status in ("missed", "dropped", "passed-no-word"):
        out["state"] = status
    else:
        out["state"] = "someday"
    return out


def progress(milestones: list[dict]) -> dict:
    goals = [m for m in milestones if m.get("source") != "expectedEnd" and m.get("status") in COUNTED]
    return {"done": sum(1 for m in goals if m.get("status") == "done"), "total": len(goals)}


def next_slug(milestones: list[dict]) -> str | None:
    """The open planned milestone with the earliest target, undated last — the
    upcoming OR overdue one, found without reading today (§6.3)."""
    open_ = [m for m in milestones if m.get("status") == "planned" and m.get("source") != "expectedEnd"]
    open_.sort(key=lambda m: (_d(m.get("target")) is None, str(m.get("target") or ""), str(m.get("slug"))))
    return open_[0]["slug"] if open_ else None


def timeline_state(state_in: dict, today) -> dict:
    """`state_in` is a wire payload's absolute fields (`input_from_timeline`
    builds it from a `ProjectTimeline`; the fixture writes it by hand)."""
    today = _d(today)
    gap = state_in.get("medianGapDays")
    if gap is None and state_in.get("momentDays"):
        gap = median_gap(state_in["momentDays"])
    q = quiet_threshold(gap)
    last = _d(state_in.get("lastMomentDay"))
    if last is None and state_in.get("momentDays"):
        last = max(_d(x) for x in state_in["momentDays"])
    idle = None if last is None else (today - last).days
    if str(state_in.get("status") or "active") in RESTING_STATUSES or idle is None or idle > QUIET_SECTION_DAYS:
        section = "resting"
    elif idle <= q:
        section = "inMotion"
    else:
        section = "quiet"
    threads = []
    for t in state_in.get("openThreads") or []:
        heard = _d(t.get("lastHeard")) or _d(t.get("since"))
        quiet_days = (today - heard).days if heard else None
        threads.append({"claimId": t.get("claimId"), "quietDays": quiet_days,
                        "followupEligible": quiet_days is not None and quiet_days >= max(FOLLOWUP_FLOOR_DAYS, q)})
    milestones = list(state_in.get("milestones") or [])
    return {
        "medianGapDays": gap, "quietThreshold": q, "section": section, "threads": threads,
        "milestones": [milestone_state(m, today) for m in milestones if m.get("source") != "expectedEnd"],
        "progress": progress(milestones), "next": next_slug(milestones), "planned": bool(milestones),
    }


def input_from_timeline(timeline) -> dict:
    """A `ProjectTimeline` (or `ProjectRow`) as `timeline_state`'s input."""
    wire = timeline.model_dump(by_alias=True)
    return {"status": (wire.get("project") or wire).get("status"), "lastMomentDay": wire.get("lastMomentDay"),
            "medianGapDays": wire.get("medianGapDays"), "momentDays": wire.get("momentDays") or [],
            "openThreads": wire.get("openThreads") or (wire.get("now") or {}).get("threads") or [],
            "milestones": wire.get("milestones") or []}
